param->ctrl translation: Fix fix_ec_paramgen_curve_nid()

levitte opened the pull request:
This function didn't prepare space to get the param string, which causes
the default_fixup_args() call to fail.

Fixes #20161

levitte committed:
param->ctrl translation: Fix fix_ec_paramgen_curve_nid()

This function didn't prepare space to get the param string, which causes
the default_fixup_args() call to fail.

Fixes #20161

levitte commented:
This is an alternative for #20226.

@ElMostafaIdrassi, would you mind testing this??

ElMostafaIdrassi commented:
@levitte I did give this a shot but it seems it it still not working. Here is the error trace:

```
80E2E94B887F0000:error:078C0102:common libcrypto routines:get_string_internal:passed a null parameter:crypto/params.c:1261:
80E2E94B887F0000:error:0A080006:SSL routines:ssl_generate_param_group:EVP lib:ssl/s3_lib.c:4811:
80E2E94B887F0000:error:0A00013A:SSL routines:tls_process_ske_ecdhe:unable to find ecdh parameters:ssl/statem/statem_clnt.c:2346:
```

I did a quick debug, and it looked like the `state == POST_PARAMS_TO_CTRL` condition is never hit, which explains why the issue still persists.

ElMostafaIdrassi commented:
I figured `state == POST_PARAMS_TO_CTRL` was just a typo, so I modified it to `state == PRE_PARAMS_TO_CTRL`. This still did not work but with a different error trace:
```
80726CF7FF7F0000:error:03000093:digital envelope routines:evp_pkey_ctx_ctrl_int:command not supported:crypto/evp/pmeth_lib.c:1327:
80726CF7FF7F0000:error:0A080006:SSL routines:ssl_generate_param_group:EVP lib:ssl/s3_lib.c:4811:
80726CF7FF7F0000:error:0A00013A:SSL routines:tls_process_ske_ecdhe:unable to find ecdh parameters:ssl/statem/statem_clnt.c:2346:
```

I finally added the missing `ctx->ctrl_cmd = EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID;` at the very end of `fix_ec_paramgen_curve_nid`. It still did not work but again with a different error trace:
```
80F26B7FE77F0000:error:08000081:elliptic curve routines:EC_GROUP_new_by_curve_name_ex:unknown group:crypto/ec/ec_curve.c:3289:name=UNDEF
80F26B7FE77F0000:error:0800008D:elliptic curve routines:pkey_ec_ctrl:invalid curve:crypto/ec/ec_pmeth.c:255:
80F26B7FE77F0000:error:0A080006:SSL routines:ssl_generate_param_group:EVP lib:ssl/s3_lib.c:4811:
80F26B7FE77F0000:error:0A00013A:SSL routines:tls_process_ske_ecdhe:unable to find ecdh parameters:ssl/statem/statem_clnt.c:2346:
```

I did a quick debug again, and the problem is in `ctx->p1 = OBJ_sn2nid(ctx->p2);`. `OBJ_sn2nid` awaits for a `const char*` but with this patch we're giving it a `ctx->p2` which is `&p2`, that is the address of a `char*`.

levitte committed:
fixup! param->ctrl translation: Fix fix_ec_paramgen_curve_nid()

levitte committed:
param->ctrl translation: Fix evp_pkey_ctx_setget_params_to_ctrl()

Ensure that ctx.ctrl_cmd defaults to translation->cmd_num

levitte commented:
> I figured `state == POST_PARAMS_TO_CTRL` was just a typo, so I modified it to `state == PRE_PARAMS_TO_CTRL`.

Yes it was a typo.  Thank you.

> I finally added the missing `ctx->ctrl_cmd = EVP_PKEY_CTRL_EC_PARAMGEN_CURVE_NID;` at the very end of `fix_ec_paramgen_curve_nid`.

It's weird that this should have been necessary...  I've made a more generic change, where I ensure that it's filled in from `translation->ctrl_num`.  See 520f9fd4f427d88abc953574e7bd53517723da04, which I just pushed here.

> I did a quick debug again, and the problem is in `ctx->p1 = OBJ_sn2nid(ctx->p2);`. `OBJ_sn2nid` awaits for a `const char*` but with this patch we're giving it a `ctx->p2` which is `&p2`, that is the address of a `char*`. Changing it to `ctx->p1 = OBJ_sn2nid(p2);` seems to work.

Ah, thank you, that was quite a blunder on my part!  Fixed that too.

ElMostafaIdrassi commented:
This seems to work, thanks!

levitte commented:
Ping?

openssl-machine commented:
This pull request is ready to merge

t8m commented:
This was merged to master, 3.1, and 3.0 branches. Closing.
